### PR TITLE
Add specific metric to time long-running `/messages` requests

### PR DESCRIPTION
**Pull request by MadLittleMods**
Add specific metric to time long-running `/messages` requests

Split out from https://github.com/matrix-org/synapse/pull/13478#discussion_r946109821

Part of https://github.com/matrix-org/synapse/issues/13356

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/rest/client/room.py [hunk 1]
```diff
@@ -60,6 +62,30 @@
 
 logger = logging.getLogger(__name__)
 
+messsages_response_timer = Histogram(
```
Comment (erikjohnston) [inline synapse/rest/client/room.py:65]: Can you add a comment to say why we're adding this on top of the existing metrics please?

File: synapse/rest/client/room.py [hunk 2]
```diff
@@ -60,6 +62,30 @@
...
+        0.25,
+        0.5,
+        1.0,
+        2.5,
+        5.0,
+        10.0,
+        30.0,
+        60.0,
+        120.0,
+        180.0,
```
Comment (erikjohnston) [inline synapse/rest/client/room.py:84]: Note: with such large buckets the resolution is going to be pretty low for anything above 10s. This is probably fine, but does mean that to move the needle meaningfully on this graph `/messages` timings would need to change by tens of seconds.

**Review by erikjohnston — COMMENTED** [review]
Otherwise LGTM

File: synapse/rest/client/room.py (hunk 2 above)
Comment (MadLittleMods) [reply synapse/rest/client/room.py:89]: I think we're good for now. Didn't want to be too greedy with the number of buckets.

Comment (erikjohnston) [reply synapse/rest/client/room.py:89]: Cool, like i said I'm way more blase of increasing the buckets on one of histograms

**Review by erikjohnston — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review and merge @erikjohnston 🐔

File: synapse/rest/client/room.py [hunk 3]
```diff
@@ -560,6 +591,7 @@ def __init__(self, hs: "HomeServer"):
         self.auth = hs.get_auth()
         self.store = hs.get_datastores().main
 
+    @messsages_response_timer.time()
```
Comment (MadLittleMods) [inline synapse/rest/client/room.py:594]: I don't think this is working how we expect. All of the buckets just have the same values which means it's saying every request is taking more than 180s.

https://prometheus.matrix.org/graph?g0.expr=synapse_room_message_list_rest_servlet_response_time_seconds_bucket&g0.tab=1&g0.stacked=0&g0.show_exemplars=0&g0.range_input=1h&g0.end_input=2022-08-18%2020%3A36%3A02&g0.moment_input=2022-08-18%2020%3A36%3A02
```
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="+Inf", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="0.005", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="0.01", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="0.025", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="0.05", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="0.1", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="0.25", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="0.5", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="1.0", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="10.0", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="120.0", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="180.0", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="2.5", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="30.0", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="5.0", service="synapse"}	79814
synapse_room_message_list_rest_servlet_response_time_seconds_bucket{environment="live", identifier="matrix.org", index="6", instance="matrix.org", job="synapse_client_reader", le="60.0", service="synapse"}	79814
```

Comment (MadLittleMods) [reply synapse/rest/client/room.py:594]: In any case, we change how this is being timed in https://github.com/matrix-org/synapse/pull/13545/files

Comment (MadLittleMods) [reply synapse/rest/client/room.py:594]: Fixed in https://github.com/matrix-org/synapse/pull/13584

File: synapse/rest/client/room.py (hunk 2 above)
Comment (MadLittleMods) [reply synapse/rest/client/room.py:89]: Increased fidelity in https://github.com/matrix-org/synapse/pull/13584#discussion_r951756543

**Pull request merged**